### ISSUE TITLE
Prefer-template prohibit multi lines

### ISSUE THREAD

**Issue by BridgeAR** [eslint/eslint#9984]
**What rule do you want to change?**

Prefer-template

**Does this change cause the rule to produce more or fewer warnings?**

More

**How will the change be implemented? (New option, new default behavior, etc.)?**

New option.

**Please provide some example code that this change will affect:**

```js
const assert = require('assert');
const foo = `multi
             line
             template
             strings
             are
             bad`;
assert(foo === 'multi line template strings are bad');
```

**What does the rule currently do for this code?**

It is happy.

**What will the rule do after it's changed?**

It should report an error that using template strings that way is not a good idea.

**Comment by LINKIWI** [comment]
This warrants some discussion; I don't see how this is "not a good idea." I can see many valid use cases for inserting newlines in template strings; this is an example:

```javascript
const ssrTemplate = (some, params)  => `
  <html>
    ${params} are used here whatever
  </html>
`;
```

**Comment by BridgeAR** [comment]
@LINKIWI that is why I suggested it as option and not as default.

**Comment by Alexsey** [comment]
I think there are two types of strings: multiline by their nature - like HTML in @LINKIWI comment and single line - like in @BridgeAR comment. Multiline template plays very well with the first type and very bad with the second type. Unfortunately, eslint cannot know what specific string is. So I think the best way to solve the problem is to add an option to allow to use + for multiline strings. In this case, the rule wouldn't force you to do things like in @BridgeAR comment but allow things like @LINKIWI comment

**Comment by BridgeAR** [comment]
@Alexsey both use cases are valid but it depends on the project. Right now the suggested option is meant for Node.js and it is probably useful in a lot of modules and projects as well. 

In case someone uses SQL or HTML or other things, it is not useful and the option should stay deactivated. But the option should not allow something, it should prohibit the use of multi line template strings.

**Comment by not-an-aardvark** [comment]
Could you elaborate on why you don't want to allow multiline template strings? I see that the assertion in the original example would fail due to the extra spaces, but I don't understand the reasoning for disallowing that syntax in general.

It's also not clear to me how this is related to the `prefer-template` rule, which only deals with string concatenation.

**Comment by BridgeAR** [comment]
One reason is that in multiple occasions people open PRs with multi line template strings that contained spaces that they were not aware of. I tried to illustrate that with my example above. Besides that it is often difficult to see where the template string begins and where it ends.

That is the reason why there is a unwritten rule in Node.js that no multi line template strings are allowed. It is currently not possible to prohibit that with eslint and new PRs are going to contain those again and people are going to complain and it produces pain on both sides: the PR creator and the reviewer.
This would not be an issue if it could simply be prohibited.

I think the `prefer-template` rule is actually the right rule to add a option for something like that. Because it often does make sense to use template strings but just not multi line ones. So the name is right and it is possible to say what type of template strings you accept.

**Comment by not-an-aardvark** [comment]
> It is currently not possible to prohibit that with eslint

There are currently several ways to prohibit that with eslint:

* Using the `no-restricted-syntax` rule:

    ```json
    {
      "rules": {
        "no-restricted-syntax": ["error", "TemplateElement[value.cooked=/\\n/]"]
      }
    }
    ```
* Using a custom rule

In general, we only add rules/options to ESLint core when they provide a clear benefit to a very significant percentage of developers. Personally, I'm not convinced that this would meet that bar, since it seems like a stylistic preference to me (although it's possible other team members would feel differently).

**Comment by BridgeAR** [comment]
I do not think that this is only about style. I definitely think that these template strings can often result in output that users did not anticipate. So even though it is possible to enforce this with the mentioned way, I think it would be best to add the option.

**Comment by BridgeAR** [comment]
As a further note: prohiting the use with either `no-restricted-syntax` or a custom rule is not possible right now due to #9983 in case `prefer-template` is used as well. Because in multiple cases `prefer-template` would actually complain that it is not used even though it is meant to be that way.

**Comment by Slayer95** [comment]
Agreed that this is not actually stylistic.

This rule would help avoid a very real footgun as it can introduce indentation into strings. For example, I step onto this recently when refactoring code, like this.

```js
let outSrc = `
${string1}
${string2}
`
`;

to
```js
let outSrc = ``;
if (someCondition) {
    outSrc = `
    ${string1}
    ${string2}
    `
} else {
    outSrc = `
    ${string3}
    ${string4}
    `
}
```

**Comment by platinumazure** [comment]
I'm with not-an-aardvark: This concern seems like a relatively narrow use case, and there are also at least two other ways to enforce this without changing the core rule. So I am :-1: to adding a change here.

**Comment by not-an-aardvark** [comment]
Thanks for your interest in improving ESLint. Unfortunately, it looks like this issue didn't get consensus from the team, so I'm closing it. We define [consensus](http://eslint.org/docs/maintainer-guide/issues.html#consensus) as having three :+1:s from team members, as well as a team member willing to champion the proposal. This is a high bar by design -- we can't realistically accept and maintain every feature request in the long term, so we only accept feature requests which are useful enough that there is consensus among the team that they're worth adding.

Since ESLint is pluggable and can load custom rules at runtime, the lack of consensus among the ESLint team doesn't need to be a blocker for you using this in your project, if you'd find it useful. It just means that you would need to [implement the rule](http://eslint.org/docs/developer-guide/working-with-rules) yourself, rather than using a bundled rule that is packaged with ESLint.
